### PR TITLE
#751 insert random participant

### LINKED ISSUE #751: Implement the insertion of a random participant
## Brief description:
"As a person that does not have a team and want to participate in HACKAUBG, I want to be able to register as a single (random) participant and leave it up to the hackathon organizers to team me up with people who are in a similar position" - Georgi Gerogiev, AUBG Student

You will write the neccessary logic that ensures the addittion of the random participant in the database after performing certain checks that ensure that there is enough space for this person to be added. 

Follow this workflow: 

image

Here is a link to the entire schema: 
https://excalidraw.com/#json=csmqdFx_8PNum3TBohYpX,AUA36WkDfzchJUbA9oBQwQ

## How to achieve it:

[ ] Create a capacity_check_one() similar to how it is done in #743 (read the description)

[ ] Use the capacity check function to check if the capacity is reached or not

[ ] Write the logic to insert the participant in the database

Keep in mind to write all the functions that manipulate the database in the repository folder either participants_repository.py or teams_repository.py and all the business logic of your task in participants_service.py.

### PR DESCRIPTION
**Pull request by KDimitrov333**
#751 insert random participant
Insert random participant:
Added capacity_check_one and the necessary logic to insert a participant
Was unable to send verification email
resolves #751

### CODE REVIEW

**Comment by Isak-Bego** [comment]
> The commit messages can be a liiiiiiittle shorter.

**Great**: "add a skeleton for adding a participant with is_admin: false"
**Too Long**: "Finish logic on capacity_check_one, add separate methods for getting number of random participants and registered teams, organise code in correct files"

**Comment by Isak-Bego** [comment]
Please write the description of the PR correctly according to the template that you are provided with and following its instructions. Use this as an example: https://github.com/AUBGTheHUB/monolith/pull/767

Something to note here that you should always put ```resolves #774``` with the number of your issue in the body of the Pull Request. This will like the issue with the pull request, and once the pull request is merged the issue will be closed with this event.

File: services/py-api/src/database/repository/participants_repository.py [hunk 1]
```diff
@@ -57,3 +58,10 @@ async def create(
         except Exception as e:
             LOG.exception("Participant insertion failed due to err {}".format(e))
             return Err(e)
+
+    async def get_number_random_participants(self):
+        return self._collection.find({"is_admin" : "false"}).count()
+        
+
+        
+
```
Comment (Isak-Bego) [inline services/py-api/src/database/repository/participants_repository.py:67]: The spaces in the end of the file should be deleted. Normally, this should be done by the pre-commit hook. Remind me to see if you have it installed in your setup or not

File: services/py-api/src/database/repository/participants_repository.py [hunk 2]
```diff
@@ -57,3 +58,10 @@ async def create(
         except Exception as e:
             LOG.exception("Participant insertion failed due to err {}".format(e))
             return Err(e)
+
+    async def get_number_random_participants(self):
+        return self._collection.find({"is_admin" : "false"}).count()
```
Comment (Isak-Bego) [inline services/py-api/src/database/repository/participants_repository.py:63]: This is now implemented in the main branch. Check 8638cbce50fa6f3458eba303a5e797c229a81b93. Pull the changes from the main branch to the branch that you are working. If you don't know how to do that just quickly google it. Resolve any conflicts if needed, you can also learn how to do that online.

File: services/py-api/src/database/repository/participants_repository.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+from math import ceil
```
Comment (Isak-Bego) [inline services/py-api/src/database/repository/participants_repository.py:1]: You won't need this here so remove it

File: services/py-api/src/database/repository/teams_repository.py [hunk 4]
```diff
@@ -57,3 +57,6 @@ async def update(
 
     async def delete(self, obj_id: str, session: Optional[AsyncIOMotorClientSession] = None) -> Result:
         raise NotImplementedError()
+    
+    async def get_number_registered_teams(self):
+        return self._collection.find({"is_verified": "true"}).count()
```
Comment (Isak-Bego) [inline services/py-api/src/database/repository/teams_repository.py:62]: Same thing with this one. This is implemented in the latest commit of the main branch. So pulling the new changes from main should give you a function which gives you the of the teams. Looks like this and its in teams_repository.py, so you can just delete the function that you have written yourself as you wont need it

```
async def get_verified_registered_teams_count(self) -> int:
        """Returns the count of verified teams."""
        # Ignoring mypy type due to Returning Any from function declared to return "int"  [no-any-return] which is not true
        return await self._collection.count_documents({"is_verified": True})  # type: ignore
```

File: services/py-api/src/server/handlers/participants_handlers.py [hunk 5]
```diff
@@ -21,7 +22,12 @@ async def create_participant(
         #     result = await self._service.register_admin_participant(input_data)
         # else:
         #     ...
-        result = await self._service.register_admin_participant(input_data)
+
+
+        if input_data.is_admin and input_data.team_name:
+            result = await self._service.register_admin_participant(input_data)
+        else:   
+            result = await self._service.register_random_participant(input_data)
```
Comment (Isak-Bego) [inline services/py-api/src/server/handlers/participants_handlers.py:30]: This is logically incorrect. If a participant is not an admin, but it has a team_name it will still be considered a random participant (although a random participant does not have a team). It should be something like: 

```Python
if input_data.is_admin:
#register the participant as admin
else if input_data.team_name:
#the participant wants to join a team through an invite link
else
#the participant is joining as a random participant
```

File: services/py-api/src/server/handlers/participants_handlers.py [hunk 6]
```diff
@@ -21,7 +22,12 @@ async def create_participant(
         #     result = await self._service.register_admin_participant(input_data)
         # else:
         #     ...
-        result = await self._service.register_admin_participant(input_data)
+
+
```
Comment (Isak-Bego) [inline services/py-api/src/server/handlers/participants_handlers.py:26]: Delete these 2 empty lines

File: services/py-api/src/server/handlers/participants_handlers.py [hunk 7]
```diff
@@ -37,3 +43,10 @@ async def create_participant(
             return ErrResponse(error="An unexpected error occurred during the creation of Participant")
 
         return ParticipantRegisteredInTeamResponse(participant=result.ok_value[0], team=result.ok_value[1])
+
+
+    async def capacity_reached(self):
+        return JSONResponse.error(
+            message = "The hackathon capacity has been reached",
+            status_code = 409
+        )
```
Comment (Isak-Bego) [inline services/py-api/src/server/handlers/participants_handlers.py:52]: Why would you have an async method to return a JSONResponse error in the handler method. If you would ever need to return a JSONResponse just use the class provided by the library no need to create a wrapper method around it. This method does not belong here, so I suggest deleting it.

File: services/py-api/src/server/handlers/participants_handlers.py [hunk 8]
```diff
@@ -1,3 +1,4 @@
+from fastapi.responses import JSONResponse
```
Comment (Isak-Bego) [inline services/py-api/src/server/handlers/participants_handlers.py:1]: You won't need this import. See the comment below.

File: services/py-api/src/server/handlers/utility_hanlders.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 from fastapi import Response, status
 
 from src.database.db_manager import DB_MANAGER
-from src.server.schemas.response_schemas.schemas import ErrResponse, PongResponse
+from src.server.schemas.response_schemas.schemas import ErrResponse, PongResponse, NumberOfRandomParticipantsResponse
```
Comment (Isak-Bego) [inline services/py-api/src/server/handlers/utility_hanlders.py:4]: No this is for testing purposes. Please make sure to clear it out before you commit!

File: services/py-api/src/server/routes/utility_routes.py [hunk 10]
```diff
@@ -19,3 +19,9 @@ async def ping(
     response: Response, handler: UtilityHandlers = Depends(create_utility_handler)
 ) -> PongResponse | ErrResponse:
     return await handler.ping_services(response)
+
+@utility_router.get("/random_users", responses={200: {"model": NumberOfRandomParticipantsResponse}, 503: {"model": ErrResponse}})
+async def random_users(
+    response: Response, handler: UtilityHandlers = Depends(create_utility_handler)
+) -> NumberOfRandomParticipantsResponse | ErrResponse:
+    return await handler.random_users_services(response)
```
Comment (Isak-Bego) [inline services/py-api/src/server/routes/utility_routes.py:27]: I may understand the purpose of it again, for testing however why would you include it in the commit. Please remove these.

File: services/py-api/src/server/routes/utility_routes.py [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 from src.database.db_manager import DB_MANAGER
 from src.server.handlers.utility_hanlders import UtilityHandlers
-from src.server.schemas.response_schemas.schemas import PongResponse, ErrResponse
+from src.server.schemas.response_schemas.schemas import PongResponse, ErrResponse, NumberOfRandomParticipantsResponse
```
Comment (Isak-Bego) [inline services/py-api/src/server/routes/utility_routes.py:5]: Remove the last import NumberOfRandomParticipantsResponse

File: services/py-api/src/server/schemas/response_schemas/schemas.py [hunk 12]
```diff
@@ -31,3 +31,7 @@ def serialize_participant(self, participant: Participant) -> Dict[str, Any]:
     @field_serializer("team")
     def serialize_team(self, team: Team) -> Dict[str, Any]:
         return team.dump_as_json()
+
+
+class NumberOfRandomParticipantsResponse(BaseModel):
+    number: int
```
Comment (Isak-Bego) [inline services/py-api/src/server/schemas/response_schemas/schemas.py:37]: Don't commit these things for the same reason as you should not commit and push the above changes.

File: services/py-api/src/service/participants_service.py [hunk 13]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
...
+        return Ok(participant.ok_value)
 
     async def register_admin_participant(
         self, input_data: ParticipantRequestBody
     ) -> Result[Tuple[Participant, Team], DuplicateEmailError | DuplicateTeamNameError | Exception]:
         # TODO: Add Capacity check 2
         result = await self._tx_manager.with_transaction(self._create_participant_and_team_in_transaction, input_data)
         return result
+    
+    async def capacity_check_one(self):
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:57]: Rename to _check_capacity_register_random_participant_case

File: services/py-api/src/service/participants_service.py [hunk 14]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
...
+        random_participants = ParticipantsRepository.get_number_random_participants()
+        registered_teams = TeamsRepository.get_number_registered_teams()
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:59]: Use the new methods after pulling the new changes from the main branch.

File: services/py-api/src/service/participants_service.py [hunk 15]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
...
+        
+        hackathon_teams_cap = 20 #change later
+        participants_per_team = 5 #change later
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:62]: You can find these in teams_repository.py as constant if you pull the latest changes. So please do so and update your code accordingly

File: services/py-api/src/service/participants_service.py [hunk 16]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
...
+
+        anticipated_total_teams = registered_teams + ceil((random_participants + 1) / participants_per_team)
+
+        if anticipated_total_teams <= hackathon_teams_cap:
+            return True
+        else:
+            return False
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:69]: This can be simplified to

```Python
return anticipated_total_teams <= hackathon_teams_cap
# The condition will be evaluated before returning and thus will give true or false. No need for that redundant logic
```

File: services/py-api/src/service/participants_service.py [hunk 17]
```diff
@@ -10,6 +11,7 @@
 from src.database.transaction_manager import TransactionManager
 from src.server.exception import DuplicateEmailError, DuplicateTeamNameError
 from src.server.schemas.request_schemas.schemas import ParticipantRequestBody
+from src.server.handlers.participants_handlers import ParticipantHandlers
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:14]: Because the service layer is called by the handlers layer, by importing the ParticipantHandler you would cause the error of circular import. Learn more about it here: https://www.geeksforgeeks.org/python-circular-imports/. Please remove this import, you should not need it if your code is designed correctly

Moreover, this is also the reason that the tests are currently not passing.

File: services/py-api/src/service/participants_service.py [hunk 18]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
...
+
+    #revise
+    async def insert_random_participant(self, input_data: ParticipantRequestBody
+    ) -> Ok[Tuple[Participant, Team]] | Err[DuplicateEmailError | DuplicateTeamNameError | Exception]:
+        
+        if ParticipantsRepository.capacity_check_one():
+           result = self._tx_manager.with_transaction(self._create_random_participant_in_transaction, input_data)
+           return result; 
+        else:
+           ParticipantHandlers.capacity_reached()
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:79]: Use something similar to this: 

```Python
async def register_admin_participant(self, input_data: ParticipantRequestBody) -> Result[
        Tuple[Participant, Team],
        DuplicateEmailError | DuplicateTeamNameError | HackathonCapacityExceededError | Exception,
    ]:

        # Capacity Check 2
        has_capacity = await self._check_capacity_register_admin_participant_case()
        if not has_capacity:
            return Err(HackathonCapacityExceededError(self._team_repo.MAX_NUMBER_OF_TEAMS_IN_HACKATHON))

        # Proceed with registration if there is capacity
        result = await self._tx_manager.with_transaction(self._create_participant_and_team_in_transaction, input_data)
        return result
```

With the new changes in main we have created a costum error type HackathonCapacityExceededError that you can throw when the check is not successful.

File: services/py-api/src/service/participants_service.py [hunk 19]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
...
+        if anticipated_total_teams <= hackathon_teams_cap:
+            return True
+        else:
+            return False
+
+    #revise
+    async def insert_random_participant(self, input_data: ParticipantRequestBody
+    ) -> Ok[Tuple[Participant, Team]] | Err[DuplicateEmailError | DuplicateTeamNameError | Exception]:
+        
+        if ParticipantsRepository.capacity_check_one():
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:75]: Why are you calling the capacity check from the Repository Layer? This method should be located in the service. See comment below

File: services/py-api/src/service/participants_service.py [hunk 20]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
             return participant
 
         return Ok((participant.ok_value, team.ok_value))
+    
+    async def _create_random_participant_in_transaction(
+            self, input_data: ParticipantRequestBody, session: Optional[AsyncIOMotorClientSession] = None
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:42]: Change the function signature to include Result[...]. Like it is done: 
```Python
async def _create_participant_and_team_in_transaction(
        self, input_data: ParticipantRequestBody, session: Optional[AsyncIOMotorClientSession] = None
    ) -> Result[Tuple[Participant, Team], DuplicateEmailError | DuplicateTeamNameError | Exception]:
```

File: services/py-api/src/service/participants_service.py [hunk 21]
```diff
@@ -35,10 +37,45 @@ async def _create_participant_and_team_in_transaction(
...
+    ) -> Ok[Participant] | Err[DuplicateEmailError | DuplicateTeamNameError | Exception]:
+        participant = await self._participant_repo.create(input_data, session, team_id=None)
```
Comment (Isak-Bego) [inline services/py-api/src/service/participants_service.py:44]: I think you can omit team_id from the arguments that you pass to create.

**Review by Isak-Bego — CHANGES_REQUESTED** [review]
Please be careful with your commits, do not include changes that are for testing purposes. Make sure to update your local branches with changes that are happening in the remote branches. Just to make sure pull from main branch before you push your changes. Thank you for taking time to work on this! We are slowly getting there!

File: services/py-api/src/server/handlers/participants_handlers.py (hunk 7 above)
Comment (IvanObreshkov) [reply services/py-api/src/server/handlers/participants_handlers.py:52]: Also we never return JSONresponse objects. Haven't you looked at the `participants_hablder.py`?

File: services/py-api/src/database/db_manager.py
Comment (IvanObreshkov) [inline services/py-api/src/database/db_manager.py:1]: Why is this edited? I know it's a new line , but pls check what you are committing. As an exercise this file should be reverted to its previous state and should not be added to this pr.

File: services/py-api/src/service/participants_service.py (hunk 19 above)
Comment (IvanObreshkov) [reply services/py-api/src/service/participants_service.py:75]: How is this even working? You need to create an instance of this class to call it's methods.

File: services/py-api/src/service/participants_service.py [hunk 22]
```diff
@@ -43,15 +43,11 @@ async def _create_random_participant_in_transaction(
...
-        if is_err(team):
-            return team
-
-        participant = await self._participant_repo.create(input_data, session, team_id=team.ok_value.id)
+        participant = await self._participant_repo.create(input_data, session, team_id=None)
         if is_err(participant):
             return participant
 
-        return Ok((participant.ok_value, team.ok_value))
+        return Ok((participant.ok_value, ""))
```
Comment (KDimitrov333) [inline services/py-api/src/service/participants_service.py:50]: I am not sure on what to return here since the random participant shouldn't have a team when he is initially registered

File: services/py-api/src/database/repository/participants_repository.py (hunk 2 above)
Comment (KDimitrov333) [reply services/py-api/src/database/repository/participants_repository.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/database/repository/participants_repository.py (hunk 3 above)
Comment (KDimitrov333) [reply services/py-api/src/database/repository/participants_repository.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/server/handlers/participants_handlers.py (hunk 6 above)
Comment (KDimitrov333) [reply services/py-api/src/server/handlers/participants_handlers.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/server/handlers/participants_handlers.py (hunk 7 above)
Comment (KDimitrov333) [reply services/py-api/src/server/handlers/participants_handlers.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/server/handlers/participants_handlers.py (hunk 8 above)
Comment (KDimitrov333) [reply services/py-api/src/server/handlers/participants_handlers.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/server/handlers/utility_hanlders.py (hunk 9 above)
Comment (KDimitrov333) [reply services/py-api/src/server/handlers/utility_hanlders.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/server/routes/utility_routes.py (hunk 10 above)
Comment (KDimitrov333) [reply services/py-api/src/server/routes/utility_routes.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/server/routes/utility_routes.py (hunk 11 above)
Comment (KDimitrov333) [reply services/py-api/src/server/routes/utility_routes.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/server/schemas/response_schemas/schemas.py (hunk 12 above)
Comment (KDimitrov333) [reply services/py-api/src/server/schemas/response_schemas/schemas.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 14 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 15 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 16 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 17 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 18 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 19 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: > Why are you calling the capacity check from the Repository Layer? This method should be located in the service. See comment below

Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 20 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: Fixed in 1bec98abe406eea6382a5453ac98f3e900d3ddd4

File: services/py-api/src/service/participants_service.py (hunk 21 above)
Comment (KDimitrov333) [reply services/py-api/src/service/participants_service.py]: Fixed in 833a26f627a807eaac502e53028d46fe1ca15419

File: services/py-api/src/server/handlers/participants_handlers.py (hunk 5 above)
Comment (IvanObreshkov) [reply services/py-api/src/server/handlers/participants_handlers.py]: Fixed in 988821a7cdf6e6755270c0cae57e8480e6456db2, so I am resolving this.

**Comment by IvanObreshkov** [comment]
@Isak-Bego I helped @KDimitrov333 with his PR. We added additional checks for edge cases. I also added a video about Dependency Injection. You can check my changes in the latest commits. Whats left is adding test which @KDimitrov333 said is going to add in this PR.

File: services/py-api/src/database/repository/teams_repository.py [hunk 23]
```diff
@@ -29,6 +29,9 @@ async def create(
         **kwargs: Dict[str, Any]
     ) -> Result[Team, DuplicateTeamNameError | Exception]:
 
+        if input_data.team_name is None:
```
Comment (asynchroza) [inline services/py-api/src/database/repository/teams_repository.py:32]: If you're using types, how come `team_name` can be null?

Comment (iobreshkov-qb) [reply services/py-api/src/database/repository/teams_repository.py:32]: it was because of mypy. I decided to do it like that, but we could also put a comment to ignore mypy err. Look at the input data schema and the Team obj model

Comment (asynchroza) [reply services/py-api/src/database/repository/teams_repository.py:32]: I'm not sure why you would be accepting something which is incorrect in the first place. Is this only for the sake of reusing the Team model?

File: services/py-api/tests/unit_tests/test_handlers/test_participant_handlers.py [hunk 24]
```diff
@@ -138,3 +137,101 @@ async def test_create_participant_capacity_exceeded_error(
...
+    participant_registration_service_mock.register_random_participant.return_value = Ok(
+        (
+            Participant(
+            name="Test User",
+            email="test@example.com",
+            is_admin=False,
+            team_id=None
+            ),
+            None)
+    )
```
Comment (IvanObreshkov) [inline services/py-api/tests/unit_tests/test_handlers/test_participant_handlers.py:159]: strange formatting

File: services/py-api/tests/unit_tests/test_database_layer/test_participant_repository.py [hunk 25]
```diff
@@ -69,3 +69,53 @@ async def test_create_participant_general_exception(
...
+    assert start_time <= result.ok_value.updated_at <= end_time, "updated_at is not within the 10-second window"
+
+@pytest.mark.asyncio
+async def test_create_participant_random_case_duplicate_email_error(
+    db_manager_mock: Mock, mock_input_data_random: ParticipantRequestBody, repo: ParticipantsRepository
+) -> None:
+    # Simulate a DuplicateKeyError raised by insert_one to represent a duplicate email
+    db_manager_mock.get_collection.return_value.insert_one = AsyncMock(
+        side_effect=DuplicateKeyError("Duplicate email error")
+    )
```
Comment (Isak-Bego) [inline services/py-api/tests/unit_tests/test_database_layer/test_participant_repository.py:100]: As we already have written tests for the repository methods. I don't think a special case for the random participant is needed. The respository methods only communciate with the database to do create, read, update operations, they should not care if we are adding a random or admin participant. Unless you write a new method that you need in the respoitory layer, you don't need to create additional test cases for it. Thank you so much for your work man!

File: services/py-api/tests/unit_tests/test_handlers/test_participant_handlers.py (hunk 24 above)
Comment (Isak-Bego) [inline services/py-api/tests/unit_tests/test_handlers/test_participant_handlers.py:159]: Format it as in test_create_particiapant_admin_success()

File: services/py-api/tests/unit_tests/test_service_layer/test_participant_registration_service.py [hunk 26]
```diff
@@ -170,3 +170,124 @@ async def test_register_admin_participant_order_of_operations(
...
+    # Mock successful `create` responses for team and participant. These are the operations inside the passed callback
+    # to with_transaction
+    participant_repo_mock.create.return_value = Participant(
+        name=mock_input_data_random.name,
+        email=mock_input_data_random.email,
+        is_admin=False,
+        team_id=None,
+    )
+
+    hackathon_service_mock.create_random_participant.return_value = Ok(participant_repo_mock.create.return_value)
```
Comment (IvanObreshkov) [inline services/py-api/tests/unit_tests/test_service_layer/test_participant_registration_service.py:193]: Doesn't `create_random_participant` return `Result[Tuple[Participant, None], DuplicateEmailError | Exception]` and not a single OK value? Maybe change this to `Ok(
        (participant_repo_mock.create.return_value, None)
    )

**Review by Isak-Bego — COMMENTED** [review]
TOP JOB

**Review by IvanObreshkov — COMMENTED** [review]
Just a few nits, but otherwise fantastic job!!! 🚀 🚀 🚀 🚀 🚀

File: services/py-api/src/database/repository/teams_repository.py (hunk 23 above)
Comment (Isak-Bego) [reply services/py-api/src/database/repository/teams_repository.py:32]: Hey Misho! We are going to keep this in mind for later on. For now we should advance in a slightly faster pace

File: services/py-api/src/service/hackathon_service.py [hunk 27]
```diff
@@ -53,6 +53,20 @@ async def create_participant_and_team_in_transaction(
             self._create_participant_and_team_in_transaction_callback, input_data
         )
 
+    async def create_random_participant(
```
Comment (asynchroza) [inline services/py-api/src/service/hackathon_service.py:56]: What's the purpose of this method? It's doing literally the same thing `self._participant_repo.create` does.

File: services/py-api/src/service/hackathon_service.py [hunk 28]
```diff
@@ -53,6 +53,20 @@ async def create_participant_and_team_in_transaction(
...
+        self, input_data: ParticipantRequestBody
+    ) -> Result[Tuple[Participant, None], DuplicateEmailError | Exception]:
+        """Creates a random participant. The participant is created with team_id=None as we create the Random teams
```
Comment (asynchroza) [inline services/py-api/src/service/hackathon_service.py:59]: This comment is incorrect. It's created with `team_id=None` only if the user passes it as such. Either fix the type or explicitly set the `team_id=None` if you decide not to ditch the method as a whole .

File: services/py-api/src/service/hackathon_service.py [hunk 29]
```diff
@@ -70,3 +84,22 @@ async def check_capacity_register_admin_participant_case(self) -> bool:
...
+        participant workflow'"""
+
+        # Fetch number of verified random participants
+        verified_random_participants = await self._participant_repo.get_verified_random_participants_count()
+
+        # Fetch number of verified registered teams
+        verified_registered_teams = await self._team_repo.get_verified_registered_teams_count()
+
+        # Calculate the anticipated number of teams if a new random participant is added
+        number_ant_teams = verified_registered_teams + ceil(
```
Comment (asynchroza) [inline services/py-api/src/service/hackathon_service.py:100]: Correct me if I'm mistaken but the logic here is incorrect. `ceil` rounds up meaning that you're accounting for an additional team which is not fully populated.

File: services/py-api/src/service/hackathon_service.py [hunk 30]
```diff
@@ -70,3 +84,22 @@ async def check_capacity_register_admin_participant_case(self) -> bool:
...
+            (verified_random_participants + 1) / self._team_repo.MAX_NUMBER_OF_TEAM_MEMBERS
+        )
+
+        # Check against the hackathon capacity
+        return number_ant_teams <= self._team_repo.MAX_NUMBER_OF_VERIFIED_TEAMS_IN_HACKATHON
```
[the remark this answers is not in the document]
Comment (asynchroza) [reply services/py-api/src/service/hackathon_service.py:105]: Never mind. Logic is good.

File: services/py-api/src/service/hackathon_service.py (hunk 29 above)
Comment (asynchroza) [reply services/py-api/src/service/hackathon_service.py:100]: Never mind. Logic is good.

File: services/py-api/src/service/hackathon_service.py (hunk 28 above)
Comment (IvanObreshkov) [reply services/py-api/src/service/hackathon_service.py:59]: So if we leave the method we should do something like this:
```python
    async def create_random_participant(
        self, input_data: ParticipantRequestBody
    ) -> Result[Tuple[Participant, None], DuplicateEmailError | Exception]:
        """Creates a random participant. The participant is created with team_id=None as we create the Random teams
        after the registration is closed, until then random participants are not added to a team."""

        # We create a copy of the input data in order to avoid modifying it directly during runtime, as this
        # could create strange behaviour down the line.
        random_participant_data = input_data.model_copy(update={"team_name": None})
        result = await self._participant_repo.create(random_participant_data)

        if is_err(result):
            return result

        # As when first created, the random participant is not assigned to a team we return the team as None
        return Ok((result.ok_value, None))
 ```
 In order to be 100% sure that `team_name` is None, no matter what is passed to it?

Comment (Isak-Bego) [reply services/py-api/src/service/hackathon_service.py:59]: Guys remember that to access this function it passes through a check that checks if the team_id=None. Thus we dont need to assign it.

Comment (Isak-Bego) [reply services/py-api/src/service/hackathon_service.py:59]:
```Python
if input_data.is_admin and input_data.team_name:
            result = await self._service.register_admin_participant(input_data)

        # TODO: Implement this when the invite participant case is done
        # elif input_data.is_admin is False and input_data.team_name:
        #     result = await self._service.register_invite_participant(input_data)

        else:
            result = await self._service.register_random_participant(input_data)
```

Comment (Isak-Bego) [reply services/py-api/src/service/hackathon_service.py:59]: Moreover
image

Comment (Isak-Bego) [reply services/py-api/src/service/hackathon_service.py:59]: @KDimitrov333 You can roll back to the previous implementation on this one. 

You only need:
```Python
result = await self._participant_repo.create(input_data)
```

File: services/py-api/src/service/hackathon_service.py [hunk 31]
```diff
@@ -59,7 +59,10 @@ async def create_random_participant(
         """Creates a random participant. The participant is created with team_id=None as we create the Random teams
```
Comment (Isak-Bego) [inline services/py-api/src/service/hackathon_service.py:59]: @KDimitrov333 remove this comment to as it creates confusion

**Review by Isak-Bego — COMMENTED** [review]
Last nits I promise hahahah. Awesome job! Please do not resolve the conversations yourself. Put Fixed in /*commit hash*/ under the comment as we will resolve them.

File: services/py-api/src/service/hackathon_service.py (hunk 28 above)
Comment (KDimitrov333) [reply services/py-api/src/service/hackathon_service.py:59]: So should I revert it to how it was before commit 1abd3aa (my last commit)?

Comment (Isak-Bego) [reply services/py-api/src/service/hackathon_service.py:59]: Yes please!

**Review by Isak-Bego — APPROVED** [review]
LGTM! Thank you Kalo! Excellent Job!
